relayer/chains/cosmos/cosmos_chain_processor.go

```diff
@@ -40,9 +60,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
 // ChainProcessors should obey the context and return upon context cancellation.
 func (ccp *CosmosChainProcessor) Run(ctx context.Context, initialBlockHistory uint64) error {
-	panic("not yet implemented")
+	// this will be used for persistence across query cycle loop executions
+	persistence := queryCyclePersistence{
+		minQueryLoopDuration: defaultMinQueryLoopDuration,
+	}
+
+	// Infinite retry to get initial latest height
+	for {
+		latestHeight, err := ccp.latestHeightWithRetry(ctx)
+		if err != nil {
+			ccp.log.Error(
+				"Failed to query latest height after max attempts",
+				zap.String("chain_id", ccp.chainProvider.ChainId()),
+				zap.Uint("attempts", latestHeightQueryRetries),
+				zap.Error(err),
+			)
+			continue
+		}
+		persistence.latestHeight = latestHeight
+		break
+	}
+
+	// this will make initial QueryLoop iteration look back initialBlockHistory blocks in history
+	latestQueriedBlock := persistence.latestHeight - int64(initialBlockHistory)
+
+	if latestQueriedBlock < 0 {
+		latestQueriedBlock = 0
+	}
+
+	persistence.latestQueriedBlock = latestQueriedBlock
+
+	ccp.log.Info("Entering main query loop", zap.String("chain_id", ccp.chainProvider.ChainId()))
+
+	ticker := time.NewTicker(persistence.minQueryLoopDuration)
+
+	for {
+		if err := ccp.queryCycle(ctx, &persistence); err != nil {
+			return err
+		}
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			ticker.Reset(persistence.minQueryLoopDuration)
+		}
+	}
+}
+
+func (ccp *CosmosChainProcessor) queryCycle(ctx context.Context, persistence *queryCyclePersistence) error {
+	chainID := ccp.chainProvider.ChainId()
+
+	var err error
+	persistence.latestHeight, err = ccp.latestHeightWithRetry(ctx)
+
+	// don't want to cause CosmosChainProcessor to quit here, can retry again next cycle.
+	if err != nil {
+		ccp.log.Error(
+			"Failed to query latest height after max attempts",
+			zap.String("chain_id", ccp.chainProvider.ChainId()),
+			zap.Uint("attempts", latestHeightQueryRetries),
+			zap.Error(err),
+		)
+		return nil
+	}
+
+	ccp.log.Debug("Queried latest height",
+		zap.String("chain_id", chainID),
+		zap.Int64("latest_height", persistence.latestHeight),
+	)
+
+	// used at the end of the cycle to send signal to path processors to start processing if both chains are in sync and no new messages came in this cycle
+	firstTimeInSync := false
+
+	if !ccp.inSync {
+		if (persistence.latestHeight - persistence.latestQueriedBlock) < inSyncNumBlocksThreshold {
+			ccp.inSync = true
+			firstTimeInSync = true
+			ccp.log.Info("chain is in sync", zap.String("chain_id", chainID))
+		} else {
+			ccp.log.Info("chain is not yet in sync",
+				zap.String("chain_id", chainID),
+				zap.Int64("latest_queried_block", persistence.latestQueriedBlock),
+				zap.Int64("latest_height", persistence.latestHeight),
+			)
+		}
+	}
+
+	for i := persistence.latestQueriedBlock + 1; i <= persistence.latestHeight; i++ {
+		blockRes, err := ccp.cc.Client.BlockResults(ctx, &i)
```

Comment:
`BlockSearch` could definitely be helpful in the case of the initial flush if we wanted to avoid using the existing flush strategy. For typical operation after the initial flush, this will be scheduled so that `persistence.latestHeight - persistence.latestQueriedBlock` will be either `0` or `1`